File: frontend/src/lib/api/secrets.ts
import {
  ChangeVaultPassword,
  GetSecretsBackend,
  InitVault,
  IsVaultRequiredForRemote,
  LockVault,
  UnlockVault,
  VaultInitialized,
  VaultUnlocked,
} from '../../../wailsjs/go/wails/SecretsService'
import { mapWailsError } from './errors'

async function wrap<T>(fn: () => Promise<T>): Promise<T> {
  try {
    return await fn()
  } catch (err) {
    throw mapWailsError(err)
  }
}

export const secretsApi = {
  getBackend: () => wrap(() => GetSecretsBackend()),
  vaultInitialized: () => wrap(() => VaultInitialized()),
  vaultUnlocked: () => wrap(() => VaultUnlocked()),
  isVaultRequiredForRemote: () => wrap(() => IsVaultRequiredForRemote()),
  initVault: (masterPassword: string) => wrap(() => InitVault(masterPassword)),
  unlockVault: (masterPassword: string) => wrap(() => UnlockVault(masterPassword)),
  lockVault: () => wrap(() => LockVault()),
  changeVaultPassword: (oldPassword: string, newPassword: string) =>
    wrap(() => ChangeVaultPassword(oldPassword, newPassword)),
}

export function isVaultLockedError(err: unknown): boolean {
  const appErr = mapWailsError(err)
  return appErr.code === 'SECRETS_VAULT_LOCKED' || appErr.code === 'SECRETS_VAULT_NOT_INITIALIZED'
}

export function isVaultWrongPasswordError(err: unknown): boolean {
  return mapWailsError(err).code === 'SECRETS_VAULT_WRONG_PASSWORD'
}

File: internal/wails/secrets.go
package wails

import (
	"github.com/wzhejunqiu/data-nexus/internal/model"
	"github.com/wzhejunqiu/data-nexus/internal/secrets"
	"github.com/wzhejunqiu/data-nexus/internal/service"
	"go.uber.org/zap"
)

type SecretsService struct {
	mgr *service.ConnectionManager
	log *zap.Logger
}

func NewSecretsService(mgr *service.ConnectionManager, log *zap.Logger) *SecretsService {
	return &SecretsService{mgr: mgr, log: log}
}

func (s *SecretsService) store() secrets.Store {
	return s.mgr.SecretsStore()
}

func (s *SecretsService) GetSecretsBackend() (string, error) {
	return call(s.log, "SecretsService.GetSecretsBackend", func() (string, error) {
		return string(s.store().ActiveBackend()), nil
	})
}

func (s *SecretsService) VaultInitialized() (bool, error) {
	return call(s.log, "SecretsService.VaultInitialized", func() (bool, error) {
		return s.store().VaultInitialized(), nil
	})
}

func (s *SecretsService) VaultUnlocked() (bool, error) {
	return call(s.log, "SecretsService.VaultUnlocked", func() (bool, error) {
		return s.store().VaultUnlocked(), nil
	})
}

func (s *SecretsService) InitVault(masterPassword string) error {
	err := callVoid(s.log, "SecretsService.InitVault", func() error {
		return s.store().InitVault(masterPassword)
	})
	if err == nil {
		s.log.Debug("vault initialized")
	}
	return err
}

func (s *SecretsService) UnlockVault(masterPassword string) error {
	err := callVoid(s.log, "SecretsService.UnlockVault", func() error {
		return s.store().UnlockVault(masterPassword)
	})
	if err == nil {
		s.log.Debug("vault unlocked")
	}
	return err
}

func (s *SecretsService) LockVault() error {
	err := callVoid(s.log, "SecretsService.LockVault", func() error {
		return s.store().LockVault()
	})
	if err == nil {
		s.log.Debug("vault locked")
	}
	return err
}

func (s *SecretsService) ChangeVaultPassword(oldPassword, newPassword string) error {
	err := callVoid(s.log, "SecretsService.ChangeVaultPassword", func() error {
		return s.store().ChangeVaultPassword(oldPassword, newPassword)
	})
	if err == nil {
		s.log.Debug("vault password changed")
	}
	return err
}

func (s *SecretsService) IsVaultRequiredForRemote() (bool, error) {
	return call(s.log, "SecretsService.IsVaultRequiredForRemote", func() (bool, error) {
		return s.store().ActiveBackend() == secrets.BackendVault, nil
	})
}

// Ensure secrets errors are typed for frontend mapping.
var (
	_ = model.ErrSecretsVaultLocked
	_ = model.ErrSecretsVaultNotInitialized
	_ = model.ErrSecretsVaultWrongPassword
)
